src/mongo/jobs/DiscoverOutdatedComposites.class.php

```diff
@@ -0,0 +1,296 @@
+     */
+    public function createJob(Array $data, $queueName=null)
+    {
+        if(!$queueName)
+        {
+            $queueName = Config::getDiscoverQueueName();
+        }
+        elseif(strpos($queueName, \Tripod\Mongo\Config::getDiscoverQueueName()) === false)
+        {
+            $queueName = \Tripod\Mongo\Config::getDiscoverQueueName() . '::' . $queueName;
+        }
+        $this->submitJob($queueName,get_class($this),$data);
+    }
+
+    /**
+     * Validate arguments in the job message
+     * @return array
+     */
+    protected function getMandatoryArgs()
+    {
+        return array(
+            self::TRIPOD_CONFIG_KEY,
+            self::STORE_NAME_KEY,
+            self::CURSOR_LIMIT
+        );
+    }
+
+    
+    public function runRegenerationTasks($regenTasks) {
+        foreach ($regenTasks as $regenTask) {
+            $specification = $regenTask->specification;
+            $compositeCollection = $regenTask->compositeCollection;
+            $compositeRegenFunction = $regenTask->compositeRegenFunction;
+            
+            foreach($regenTask->cbdDocuments as $cbdDoc) {
+                $compositeRegenFunction($specification, $compositeCollection, $cbdDoc);
+            }
+        }
+    }
+
+    /**
+     * Generates regeneration task data for any outdated documents in a composite collection,
+     * by scanning that composite collection one cursor-page at a time.
+     *
+     * @param CompositeMetadata $metadatum - the composite collection to search
+     * @return CompositeRegenTask|null - a CompositeRegenTask, if outdated documents are found; null otherwise
+    **/
+    public function getRegenTaskForMetadata($metadatum, $cursorLimit) {
+        $filterOutdated = $metadatum->getOutdatedQueryComponent();
+        if(!isset($filterOutdated)) {
+            return null;
+        } else {
+            $compositeCollection = $metadatum->compositeCollection;
+            $specification = $metadatum->specification;
+            $cbdCollection = $metadatum->cbdCollection;
+            $compositeRegenFunction = $metadatum->compositeRegenFunction;
+
+            $cbdIdFields = array(_ID_KEY.'.'._ID_RESOURCE => 1, _ID_KEY.'.'._ID_CONTEXT => 1);
+
+            // find one cursor page at a time for this composite
+            $outdatedComposites = 
+                $compositeCollection
+                    ->find($filterOutdated, $cbdIdFields)
+                    ->limit($cursorLimit);
+
+            if(!$outdatedComposites->hasNext()) {
+                // no outdated composite documents found, so no work to be done
+                return null;
+            } else {
+                // find IDs root CBDs required to regenerate the composite
+                $outdatedCbdIds = 
+                    array_map(array($this, 'composite2cbdId'), iterator_to_array($outdatedComposites, false));
+
+                // fetch the CBDs themselves
+                $filterCbdsById = array('$or' => $outdatedCbdIds);
```

Comment:
Is it possible to starve certain composite types, if a frequently-updated version is always before it in the `$or`?